rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,117 @@
+
+With the templated terraform, minaform can now use the `coda-network` tool (aka carey's key service), to look for a set of keys and genesis block in google cloud for this testnet name, and if they do not exist then it will generate all of the neccessary keys for you, as well as the ledger itself.
+
+Flags to the `create-network` command for `--generate-keys`, `--upload-keys`, and `--download-keys` allow you to force regeneration/re-upload of keys for easier testing and quick iteration on the key generation tooling. The default value of `true` for each of these makes it easy to share keys across O(1)  Labs employees who are collaborating on the same network (Pain Point #6).
+
+Additional flags can be provided to the create-network command for changing the quantities of fish, whales, seeds, bots, etc.
+
+The flag `--generate-image` will cause the tool to build a new image based on the provided one, but with the genesis ledger, proofs, and keys baked-in to reduce network startup time at the expense of this one-time step (Pain Point #4).
+
+Two flags are availible to specify which version of the daemon to run, `--daemon-image` and `--daemon-git-sha` where daemon-git-sha calculates the image tag based on the given git tag (Pain Point #5). 
+
+Once the network has been defined and set up, you can make manual changes to the terraform files to customize the new network, or just proceed to `minaform start <testnet_name>`.
+
+### Minaform Start
+
+The start command handles deploying the network and starting all containers / gcloud resources. `minaform start <testnet_name>` will start the given testnet on the default cluster, and provide a output to display the deployment progress. Keys will be uploaded to the cluster when the namespace is created, and then terraform resources will be deployed. If `--wait` is specified, the script will run `minaform watch <testnet_name>` once terraform is complete to watch for all of the nodes to become ready.
+
+### Minaform Stop
+
+The stop command handles destroying all resources for the given testnet, ideally using the existing terraform state and all of the same pararemeters so destroys can always happen smoothly. Will provide a `--force` flag to force terraform and other tooling to delete everything and clean up. This command closely matching the process for start/restart/etc. addresses pain point #3 and should make it easier to keep the cluster clean and easy.
+
+### Minaform Restart
+
+The restart command runs a `minaform stop` followed by `minaform start` and accepts the same flags as both start + stop.
+
+### Minaform Upgrade
+
+The upgrade command runs a terraform apply, to update the existing deployment with new changes. Ideally this should also provide more debugging flags to terraform and other tooling, and require user input to proceed so that the engineer deploying can understand what state will be lost (if any) in the process. Accepts the same flags as start/stop/restart.
```

Comment:
We need to probably put more thought into how we want terraform reapplications to actually be setup, because we don't use this workflow atm.